Data/Pool.hs

```diff
@@ -349,33 +357,37 @@ tryWithResource :: forall m a b.
 tryWithResource pool act = control $ \runInIO -> mask $ \restore -> do
   res <- tryTakeResource pool
   case res of
-    Just (resource, local) -> do
+    (Just (resource, local), _isConnPoolAlmostSaturated) -> do
       ret <- restore (runInIO (Just <$> act resource)) `onException`
                 destroyResource pool local resource
       putResource local resource
       return ret
-    Nothing -> restore . runInIO $ return (Nothing :: Maybe b)
+    (Nothing, _) -> restore . runInIO $ return (Nothing :: Maybe b)
 {-# INLINABLE tryWithResource #-}
 
 -- | A non-blocking version of 'takeResource'. The 'tryTakeResource' function
 -- returns immediately, with 'Nothing' if the pool is exhausted, or @'Just' (a,
--- 'LocalPool' a)@ if a resource could be borrowed from the pool successfully.
-tryTakeResource :: Pool a -> IO (Maybe (a, LocalPool a))
+-- 'LocalPool' a, Bool)@ if a resource could be borrowed from the pool successfully.
+-- The Bool is True, if the connection pool is near saturation i.e 80% or more of 
+-- the resources are currently in use, else it is False. 
+tryTakeResource :: Pool a -> IO (Maybe (a, LocalPool a), Bool)
 tryTakeResource pool@Pool{..} = do
   local@LocalPool{..} <- getLocalPool pool
-  resource <- liftBase . join . atomically $ do
+  (resource, isConnPoolAlmostSaturated) <- liftBase . join . atomically $ do
```

Comment:
On further pondering, I think for the `almost saturated` scenario. Doing this is not an ideal approach. Rather we could just write to a TVar inside the STM block and before returning the resources just run an IO action.

Something like:
```haskell
isPoolAlmostSaturated <- newTVarIO False
resource <- liftBase . join . atomically $ do
 ...
when ((0.8 * fromIntegral maxResources) >= fromIntegral used) $ do
 writeTVar isPoolAlmostSaturated True

```